src/formats/CIF.cpp

```diff
@@ -0,0 +1,427 @@
+            auto z = cif_to_double(line_split[cartn_z->second].to_string());
+
+            frame.add_atom(std::move(atom), Vector3D(x, y, z));
+        } else {
+            auto u = cif_to_double(line_split[fract_x->second].to_string());
+            auto v = cif_to_double(line_split[fract_y->second].to_string());
+            auto w = cif_to_double(line_split[fract_z->second].to_string());
+
+            auto vector = cell_.matrix() * Vector3D(u, v, w);
+            frame.add_atom(std::move(atom), vector);
+        }
+
+        position = file_.tellpos();
+
+        if (label_comp_id == atom_site_map_.end() || label_asym_id == atom_site_map_.end()) {
+            continue;
+        }
+
+        auto atom_id = frame.size() - 1;
+        int64_t resid = 0;
+        auto resid_text = line_split[label_seq_id->second];
+
+        try {
+            if (resid_text == ".") { // In this case, we need to use the entity id
+                resid = parse<int64_t>(line_split[label_entity_id->second]);
+            } else {
+                resid = parse<int64_t>(line_split[label_seq_id->second]);
+            }
+        } catch (const Error& e) {
+            throw format_error("invalid CIF residue or entity numeric: {}", e.what());
+        }
+
+        auto chainid = line_split[label_asym_id->second].to_string();
+
+        if (residues_.find({chainid, resid}) == residues_.end()) {
+
+            auto name = line_split[label_comp_id->second];
+            Residue residue(name.to_string(), resid);
+            residue.add_atom(atom_id);
+
+            // This will be saved as a string on purpose to match MMTF
+            if (label_asym_id != atom_site_map_.end()) {
+                residue.set("chainid", chainid);
+            }
+
+            if (auth_asym_id != atom_site_map_.end()) {
+                residue.set("chainname", line_split[auth_asym_id->second].to_string());
+            }
+
+            if (group_pdb != atom_site_map_.end()) {
+                residue.set("is_standard_pdb", line_split[group_pdb->second] == "ATOM");
+            }
+
+            residues_.emplace(std::make_pair(chainid, resid), std::move(residue));
+        } else {
+            // Just add this atom to the residue
+            residues_.at({chainid, resid}).add_atom(atom_id);
+        }
+    }
+
+    // Reset state to previous line
+    file_.seekpos(position);
+
+    for (const auto& residue: residues_) {
+        frame.add_residue(residue.second);
+    }
+
+    // Only link if we are reading CIF
+    if (model_position != atom_site_map_.end()) {
+        // Cross format talk! Forgive me!
+        PDBFormat::link_standard_residue_bonds(frame);
+    }
+}
+
+void CIFFormat::write(const Frame& frame) {
+    file_.print("data_model_{}\n", models_); // TODO: do models have a name in Chemfiles? If so, use it
+    file_.print("_audit_creation_method         'generated by Chemfiles'\n");
+    file_.print("_symmetry_cell_setting         'triclinic'\n");
```

Comment:
Could this use `frame.cell.shape` instead to set the setting to orthorhombic? There is also the question of what to do with infinite cells.

Reply:
Even if the cell has equal cell vectors and angles of 90°, in crystallographic terms as long as its space group is P1 it is in the triclinic cell setting.

Reply:
Regarding infinite cells, other packages are facing the same question, and it's not clear what is the best answer (see for example https://gitlab.com/ase/ase/-/issues/286).

What Mercury does is add a fake cubic cell with 1 Å size, so fractional coordinates are equal to Cartesian:

```
_cell_length_a                   1
_cell_length_b                   1
_cell_length_c                   1
_cell_angle_alpha                90
_cell_angle_beta                 90
_cell_angle_gamma                90
```

Reply:
> Even if the cell has equal cell vectors and angles of 90°, in crystallographic terms as long as its space group is P1 it is in the triclinic cell setting.

Oh, ok that's good to know!

> What Mercury does is add a fake cubic cell with 1 Å size, so fractional coordinates are equal to Cartesian

That sounds good, I think the same convention is applied in PDB and potentially mmCIF files when no cell information is present. The issue would be when reading such file, we would like to create an infinite cell back, not a fake cubic cell. A potential criteria to distinguish fake 1A cubic cells from real ones could be the presence of atoms with coordinates of 4A or more. Although real 1A cubic cells should be quite uncommon anyway ^^